src/lib/momento/auth/credential_provider.ex

```
@@ -0,0 +1,193 @@
+defmodule Momento.Auth.CredentialProvider do
+  require Joken
+
+  @moduledoc """
+  Handles decoding and managing Momento authentication credentials.
+  """
+
+  @enforce_keys [
+    :control_endpoint,
+    :cache_endpoint,
+    :auth_token
+  ]
+  defstruct [
+    :control_endpoint,
+    :cache_endpoint,
+    :auth_token
+  ]
+
+  @opaque t :: %__MODULE__{
+            control_endpoint: String.t(),
+            cache_endpoint: String.t(),
+            auth_token: String.t()
+          }
+
+  @doc """
+  Fetches the given environment variable and parses it into a credential.
+
+  Returns the credential or raises an exception.
+
+  Supply control_endpoint or cache_endpoint in the options to override them.
+
+  ## Examples
+
+      iex> Momento.Auth.Credential.from_env_var!("MOMENTO_AUTH_TOKEN")
+      %Momento.Auth.Credential{}
+
+  """
+  @spec from_env_var!(String.t(), keyword()) :: t()
```

I want to get this merged so we can move on to get/set but I would like to discuss this at some point.

The UX for the API seems good/fine, and I'm guessing that you found that this is an idiom. My only hang-up is that it seems like this would allow a user to pass literally any options args that they wanted to, and then we would just silently ignore them if they passed something other than our two expected ones. Is that accurate?

If so, the worst case scenario is that someone makes a typo in one of those two names and we just silently ignore the arg and they can't figure out why.

As a developer in general I would much prefer to get some kind of error in that scenario, be it "compile" time or run time. Not sure if there is another way we could spec this that would enforce that. Obv we could do our own runtime checking and throw an err on unrecognized opts but that kinda stinks.

Anyway, not a blocker, just curious to get your thoughts.

---

I agree that it it makes more sense to make those two fields explicit.

---

cool. let's not block this PR on that though, this is good foundation to get in place.